components/zcash_protocol/src/value.rs

```diff
@@ -0,0 +1,512 @@
+        self > &Zatoshis::ZERO
+    }
+}
+
+impl From<Zatoshis> for ZatBalance {
+    fn from(n: Zatoshis) -> Self {
+        ZatBalance(n.0 as i64)
+    }
+}
+
+impl From<&Zatoshis> for ZatBalance {
+    fn from(n: &Zatoshis) -> Self {
+        ZatBalance(n.0 as i64)
+    }
+}
+
+impl From<Zatoshis> for u64 {
+    fn from(n: Zatoshis) -> Self {
+        n.into_u64()
+    }
+}
+
+impl TryFrom<u64> for Zatoshis {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        Zatoshis::from_u64(value)
+    }
+}
+
+impl TryFrom<ZatBalance> for Zatoshis {
+    type Error = ();
+
+    fn try_from(value: ZatBalance) -> Result<Self, Self::Error> {
+        Zatoshis::from_nonnegative_i64(value.0)
+    }
+}
+
+impl Add<Zatoshis> for Zatoshis {
+    type Output = Option<Zatoshis>;
+
+    fn add(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        Self::from_u64(self.0.checked_add(rhs.0)?).ok()
+    }
+}
+
+impl Add<Zatoshis> for Option<Zatoshis> {
+    type Output = Self;
+
+    fn add(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        self.and_then(|lhs| lhs + rhs)
+    }
+}
+
+impl Sub<Zatoshis> for Zatoshis {
+    type Output = Option<Zatoshis>;
+
+    fn sub(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        Zatoshis::from_u64(self.0.checked_sub(rhs.0)?).ok()
+    }
+}
+
+impl Sub<Zatoshis> for Option<Zatoshis> {
+    type Output = Self;
+
+    fn sub(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        self.and_then(|lhs| lhs - rhs)
+    }
+}
+
+impl Mul<usize> for Zatoshis {
+    type Output = Option<Self>;
+
+    fn mul(self, rhs: usize) -> Option<Zatoshis> {
+        Zatoshis::from_u64(self.0.checked_mul(u64::try_from(rhs).ok()?)?).ok()
+    }
+}
+
+impl Sum<Zatoshis> for Option<Zatoshis> {
+    fn sum<I: Iterator<Item = Zatoshis>>(iter: I) -> Self {
+        iter.fold(Some(Zatoshis::ZERO), |acc, a| acc? + a)
+    }
+}
+
+impl<'a> Sum<&'a Zatoshis> for Option<Zatoshis> {
+    fn sum<I: Iterator<Item = &'a Zatoshis>>(iter: I) -> Self {
+        iter.fold(Some(Zatoshis::ZERO), |acc, a| acc? + *a)
+    }
+}
+
+/// A type for balance violations in amount addition and subtraction
+/// (overflow and underflow of allowed ranges)
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum BalanceError {
+    Overflow,
+    Underflow,
+}
+
+impl error::Error for BalanceError {}
+
+impl std::fmt::Display for BalanceError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match &self {
+            BalanceError::Overflow => {
+                write!(
+                    f,
+                    "ZatBalance addition resulted in a value outside the valid range."
+                )
+            }
+            BalanceError::Underflow => write!(
+                f,
+                "ZatBalance subtraction resulted in a value outside the valid range."
+            ),
+        }
+    }
+}
+
+impl From<Infallible> for BalanceError {
+    fn from(_value: Infallible) -> Self {
+        unreachable!()
+    }
+}
```

Comment:
I was surprised this didn't exist as a blanket `impl<T> From<Infallible> for T where T: Error`. I think the reason is https://github.com/rust-lang/libs-team/issues/233#issuecomment-1572758234. (There does exist [`impl<T> From<!> for T`](https://doc.rust-lang.org/std/convert/enum.Infallible.html#impl-From%3C!%3E-for-T). It will be nice when https://github.com/rust-lang/rust/issues/35121 lands.)